starlite/kwargs.py

```diff
@@ -332,51 +334,48 @@ def to_kwargs(self, connection: Union["WebSocket", "Request"]) -> Dict[str, Any]
-                if param.is_required
-                else connection_query_params.get(param.field_alias, param.default_value)
-                for param in self.expected_query_params
-            }
-            header_params = {
-                param.field_name: connection.headers[param.field_alias]
-                if param.is_required
-                else connection.headers.get(param.field_alias, param.default_value)
-                for param in self.expected_header_params
-            }
-            cookie_params = {
-                param.field_name: connection.cookies[param.field_alias]
-                if param.is_required
-                else connection.cookies.get(param.field_alias, param.default_value)
-                for param in self.expected_cookie_params
-            }
-            return {**reserved_kwargs, **path_params, **query_params, **header_params, **cookie_params}
-        except KeyError as e:
-            raise ValidationException(f"Missing required parameter {e.args[0]} for url {connection.url}") from e
+
+        path_params = self._collect_params(
+            params=connection.path_params, expected=self.expected_path_params, url=connection.url
+        )
+        query_params = self._collect_params(
+            params=connection_query_params, expected=self.expected_query_params, url=connection.url
+        )
+        header_params = self._collect_params(
+            params=connection.headers, expected=self.expected_header_params, url=connection.url
+        )
+        cookie_params = self._collect_params(
+            params=connection.cookies, expected=self.expected_cookie_params, url=connection.url
+        )
+
+        if not self.expected_reserved_kwargs:
+            return {**path_params, **query_params, **header_params, **cookie_params}
+
+        reserved_kwargs: Dict[str, Any] = {}
+        if "state" in self.expected_reserved_kwargs:
+            reserved_kwargs["state"] = connection.app.state.copy()
+        if "headers" in self.expected_reserved_kwargs:
+            reserved_kwargs["headers"] = connection.headers
+        if "cookies" in self.expected_reserved_kwargs:
+            reserved_kwargs["cookies"] = connection.cookies
+        if "query" in self.expected_reserved_kwargs:
+            reserved_kwargs["query"] = connection_query_params
+        if "request" in self.expected_reserved_kwargs:
+            reserved_kwargs["request"] = connection
+        if "socket" in self.expected_reserved_kwargs:
+            reserved_kwargs["socket"] = connection
+        if "data" in self.expected_reserved_kwargs:
+            reserved_kwargs["data"] = self._get_request_data(request=cast("Request", connection))
+        return {**reserved_kwargs, **path_params, **query_params, **header_params, **cookie_params}
+
+    @staticmethod
+    def _collect_params(params: Mapping[str, Any], expected: Set[ParameterDefinition], url: URL) -> Dict[str, Any]:
```

Comment:
This is marginally slower than what was happening before (because there are two iterations through the required params) but has the advantage of picking up all missing params, rather than just the first.

Reply:
speed is not a concern here because the logic in this file is running once - on application startup, and pre-computing the results.